etcd/v1/types.go

```diff
@@ -0,0 +1,84 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterMemberRequest describes a request to join the cluster by a potential etcd cluster member.
+type ClusterMemberRequest struct {
+	// Standard object's metadata.
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// The membership request itself.
+	// +optional
+	Spec ClusterMemberRequestSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Derived information about the request.
+	// +optional
+	Status ClusterMemberRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ClusterMemberRequestSpec defines an individual etcd cluster member request.
+type ClusterMemberRequestSpec struct {
+	// Name of the etcd peer member requesting membership.
+	Name string `json:"Name" protobuf:"bytes,1,opt,name=name"`
+	// PeerlURLs is comma seperated string including the members PeerURLs.
+	// +optional
+	PeerURLs string `json:"PeerURLs" protobuf:"bytes,2,opt,name=peerUrls"`
+}
+
+type ClusterMemberRequestStatus struct {
+	// Conditions applied to the request, such as approval or denial.
+	// +optional
+	Conditions []ClusterMemberRequestCondition `json:"conditions,omitempty" protobuf:"bytes,1,rep,name=conditions"`
+
+	// If request was approved, the controller will place the populated config here.
+	// +optional
+	Config ClusterMemberConfig `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`
+}
+
+type RequestConditionType string
+
+// These are the possible conditions for the etcd membership request.
+const (
+	ClusterMemberApproved  RequestConditionType = "Approved"
+	ClusterMemberDelivered RequestConditionType = "Delivered"
+	ClusterMemberDenied    RequestConditionType = "Denied"
+	ClusterMemberPending   RequestConditionType = "Pending"
+)
+
+type ClusterMemberRequestCondition struct {
+	// request approval state, currently Approved, Delivered, Denied or Pending.
+	Type RequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
+	// brief reason for the request state
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,2,opt,name=reason"`
+	// human readable message with details about the request state
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
+	// timestamp for the last update to this condition
+	// +optional
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,4,opt,name=lastUpdateTime"`
+}
+
+type ClusterMemberConfig struct {
+	// Initial cluster configuration for bootstrapping. The value will be a comma delimted string
+	// containing a list of etcd peers in the format <name>=<address>.
+	InitialCluster string `json:"initialCluster,omitempty" protobuf:"bytes,2,opt,name=initialCluster"`
```

Comment:
My understanding is that the list of peer urls will change over time.  The doc here suggests that this will not be updated over time.  Am I reading it incorrectly?

Reply:
Let me try to better explain. This value is only ever used by the etcd member once during membership scaling process to add the member to the cluster. After the member is started this value would no longer be consumed as the membership details of the cluster are stored in the member bucket of the state machine. So although the list of peers could change this value is static to the request itself at the time of the request. Once the config is delivered the data here is no longer used.

Much like a CSR request, this would be a single-use transaction.